cilium/CFP-30572-non-default-deny-policies.md

```diff
@@ -0,0 +1,175 @@
+# CFP-30572: Non-default-Deny Network Policy
+
+**SIG: SIG-Policy**
+
+**Begin Design Discussion:** 2024-02-02
+
+**Cilium Release:** 1.16
+
+**Authors:** Casey Callendrello <cdc@isovalent.com>
+
+## Summary
+
+Allow users to create network policies that do not drop non-matching traffic. 
+
+## Motivation
+
+Consider a large, multitenant cluster. The cluster administrator is distinct from
+the workload owners, and the workloads are diverse and existing.
+
+The administrator may wish to create a Cilium Clusterwide Network Policy (CCNP)
+that affects most or all existing workloads. For example, they may wish to
+- Globally grant access to a security scanner
+- Ensure that workloads can access kube-dns
+- Proxy all DNS requests for monitoring purposes
+- Deny access to sensitive IPs, such as cloud provider metadata APIs
+
+All of these can already be accomplished by the existing policy language.
+
+By default, all Cilium endpoints have unrestricted network access. If at least one
+network policy applies to an endpoint, then all traffic is denied except that
+explicitly allowed by policy.
+
+This presents a challenge for the cluster administrator. They cannot assume that
+existing, heterogeneous worloads already have network policies provisioned. So, by
+creating an administrative network policy, they may select endpoints that **previously
+were in an allow-all state**. Thus, they may unintentionally block legitimate
+traffic for an existing workload, causing an outage.
+
+## Goals
+
+* Ability to safely roll-out broad-based network policies, without the risk
+  of disrupting existing traffic.
+
+## Non-Goals
+
+* An iterative, iptables-style firewall rules engine
+* Arbitrary, multi-level rule priority
+
+## Proposal
+
+### Overview
+
+We propose a new field in the policy spec, `DefaultAction`. This field specifies
+whether or not an implicit default-deny rule should be created _on behalf of this policy_.
+
+This field is optional. The default value is `Deny`. This matches the existing behavior, wherein any network policy implicitly blocks all other traffic.
+
+The value `Pass` indicates that this policy should not create a default-deny rule.
+
+Note that this is unrelated to the `IngressDeny` and `EgressDeny` policy expressions. Those dictate the action for *matching* traffic, whereas the `DefaultAction` specifies what should happen with *non-matching* traffic.
```

Comment:
I found myself falling into a bit of a trap with the wording around `non-matching`. In the newer ANP terminology they clearly define the _subject_ of the policy and separately the rules that apply to that subject. In this case, the "non-matching traffic" is all traffic that is part of the direction in the rules but not explicitly described in the rules, but also *still part of the subject*. Ie if you use an endpointSelector to apply policy for `app: foo` within namespace `foo`, then this has no impact on the policy for `app: bar` in namespace `bar`. The non-matching traffic is the traffic to/from the subject which is not explicitly described in the policy.

I think that this existing description is probably clear enough, but we might consider mentioning the subject:
```suggestion
Note that this is unrelated to the `IngressDeny` and `EgressDeny` policy expressions. Those dictate the action for *matching* traffic for the endpoint subject to the policy, whereas the `DefaultAction` specifies what should happen with *non-matching* traffic for the subject endpoint.
```

(Not entirely sure if this helps but wording is always a bit tough. Feel free to adopt, shuffle around this suggestion or reject it altogether)

Reply:
I changed the wording a bit, using the terms "subject" and "peer". This is at least vaguely consistent with the k8s terminology.